src/net.cpp

```diff
@@ -1886,8 +1899,11 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 }
                 LogPrint(BCLog::NET, "Making feeler connection to %s\n", addrConnect.ToStringAddrPort());
             }
-
-            OpenNetworkConnection(addrConnect, (int)setConnected.size() >= std::min(nMaxConnections - 1, 2), &grant, nullptr, conn_type);
+            // Record addrman failure attempts when node has at least 2 persistent outbound connections to peers with
+            // different netgroups in ipv4/ipv6 networks + all peers in Tor/I2P/CJDNS networks.
+            // Don't record addrman failure attempts when node is offline. This can be identified since all local
+            // network connections(if any) belong in the same netgroup and size of setConnected would only be 1.
+            OpenNetworkConnection(addrConnect, (int)setConnected.size() + outbound_privacy_network_peers >= std::min(nMaxConnections - 1, 2), &grant, nullptr, conn_type);
```

Comment:
This `(int)setConnected.size() >= std::min(nMaxConnections - 1, 2)` was introduced in c769c4af11fc58dd4813d328c7f71042bc577676.

I think that for the purposes of checking whether we are offline, being connected to just one Tor peer means that we are not. This is because the Tor network is global (same for I2P or CJDNS). There is no such thing as "local area network" of Tor nodes (e.g. `10.0.0.0/24`) or localhost Tor address (like `127.0.0.1` in IPv4). If we are connected to one Tor node, it means we are not offline and can possibly connect to other Tor nodes as well.

Maybe out of the scope of this PR, but would it not be better to change that logic to: "if we are connected to at least one `CNetAddr::IsRoutable()` IPv4/6 address or at least one Tor address or at least one I2P or at least one CJDNS address, then we are not offline"?

cc @gmaxwell

Reply:
In #8065, @gmaxwell pointed out:

> This is still somewhat incomplete protection because our outbound peers could be down but not timed out

We never really know if our peers are connected or not, because when your network goes down it usually takes some amount of time for each connection to time out. Requiring two connections to be up (versus just 1 connection to e.g. Tor, or some other non-local ipv4/ipv6 network) is a tradeoff around how long we spend in a state where we're punishing addrman entries while we might be offline and just haven't noticed it yet, vs not punishing entries for being down when our network is up.

My guess is that most nodes on the network don't have local network peers, and so the effect of this logic for most users since it was deployed in #8065 -- requiring 2 non-local outbound connections before we start punishing addrman entries for failed attempts -- will be the same after this PR is merged.

If someone can figure out that it's more optimal for this to be 1 or 3 or some other value, I think we can pick that up in a new PR; however I'd suspect it doesn't matter too much, and that 2 is likely better than 1...